Refactor: move RowGroupPredicateBuilder into its own module, rename to PruningPredicateBuilder

alamb opened the pull request:
# Which issue does this PR close?
Part of https://github.com/apache/arrow-datafusion/issues/363


 # Rationale for this change

As explained on https://github.com/apache/arrow-datafusion/issues/363 the high level idea goal is to make the parquet row group pruning logic generic to any types of min/max statistics (not just parquet metadata)

# What changes are included in this PR?
1. Move code out of parquet.rs and into its own module
2. Rename to StatsPredicateBuilder to reflect its purpose

No changes in functionality are intended

The PR contains two commits to help reviewers


# Are there any user-facing changes?
The `pub struct RowGroupPredicateBuilder` struct was renamed

alamb commented:
FYI @yordan-pavlov and @returnString

jorgecarleitao reviewed (approved):
I have not read it in detail but I assume it is just a code movement and rename and therefore 👍

Dandandan commented on datafusion/src/physical_optimizer/pruning.rs:
```diff
@@ -0,0 +1,769 @@
+        let logical_predicate_expr =
+            build_predicate_expression(expr, &schema, &mut stat_column_req)?;
+        // println!(
+        //     "PruningPredicateBuilder::try_new, logical_predicate_expr: {:?}",
```
Those println comments could be removed?

Dandandan replied:
(they were there in the original code but I think would be better to remove them)

Dandandan commented on datafusion/src/physical_optimizer/pruning.rs:
```diff
@@ -0,0 +1,769 @@
+        self.scalar_expr
+    }
+
+    // fn column_name(&self) -> &String {
```
Could be removed

alamb replied:
in 688dde41b748801183d3484c6116442c22bdf591

Dandandan commented on datafusion/src/physical_optimizer/pruning.rs:
```diff
@@ -0,0 +1,769 @@
+        }
+    }
+
+    // fn column_expr(&self) -> &Expr {
```
Could be removed

alamb replied:
in 688dde41b748801183d3484c6116442c22bdf591

Dandandan commented on datafusion/src/physical_optimizer/pruning.rs:
```diff
@@ -0,0 +1,769 @@
+    fn is_stat_column_missing(&self, statistics_type: StatisticsType) -> bool {
+        self.stat_column_req
+            .iter()
+            .filter(|(c, t, _f)| c == &self.column_name && t == &statistics_type)
```
This code could use `.any` instead of filter + count.

Dandandan commented:
I also think this is looking good. I think we might clean things up a bit while touching the code, I added some suggestions.

alamb commented:
> I also think this is looking good. I think we might clean things up a bit while touching the code, I added some suggestions.

Cleaned per @Dandandan 's suggestions

Dandandan reviewed (approved):
I like it, great use of the physical optimizer
